shark_turbine/kernel/wave/constraints.py

```diff
@@ -0,0 +1,62 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from ..lang import sym
+from .._support.indexing import IndexExpr
+
+
+@dataclass
```

Comment:
Why is this a data class despite having no fields?

Reply:
It does not strictly have to be a dataclass right now, but adding the descriptor ensures we don't forget to add it when introducing new fields common to all constraints in the future.
I think class hierarchies involving dataclasses are least surprising when all classes are dataclasses.
For that reason I preemptively made `Constraint` a dataclass.